### PR TITLE
Backend for Open collaboration

### PR DESCRIPTION
**Pull request by leogrim**
@andrewconner @eishay @ryanpbrewster

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeeperInfo.scala [hunk 1]
```diff
@@ -124,6 +125,7 @@ object LibraryData {
     (__ \ 'hasCollaborators).write[Boolean] and
     (__ \ 'subscribedToUpdates).write[Boolean] and
     (__ \ 'collaborators).write[Seq[BasicUser]] and
-    (__ \ 'orgAvatar).writeNullable[ImagePath]
+    (__ \ 'orgAvatar).writeNullable[ImagePath] and
+    (__ \ 'membership).writeNullable[LibraryMembershipInfo]
   )(unlift(LibraryData.unapply))
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeeperInfo.scala:130]: @andrewconner

File: kifi-backend/modules/shoebox/conf/shoeboxService.routes [hunk 2]
```diff
@@ -112,7 +112,7 @@ POST    /ext/pref/email/:kind/:send @com.keepit.controllers.ext.ExtPreferenceCon
 GET     /ext/contacts/search        @com.keepit.controllers.ext.ExtUserController.searchForContacts(query:Option[String], limit:Option[Int])
 POST    /ext/contacts/hide          @com.keepit.controllers.ext.ExtNonUserSearchController.hideEmailFromUser()
 
-GET     /ext/libraries              @com.keepit.controllers.ext.ExtLibraryController.getLibraries()
+GET     /ext/libraries              @com.keepit.controllers.ext.ExtLibraryController.getLibraries(includeOrgLibraries: Boolean ?= false)
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/conf/shoeboxService.routes:115]: @andrewconner

File: kifi-backend/modules/shoebox/marketing [hunk 3]
```diff
@@ -1 +1 @@
-Subproject commit 207f54c6f63ecc7f8144aab5acb1aefb0b4da105
+Subproject commit ea88df22df4d8e6c3e9bb8616e73ab1308fe2b39
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/marketing:1]: @andrewconner how do I get rid of this again?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryMembershipCommander.scala [hunk 4]
```diff
@@ -74,4 +115,285 @@ class LibraryMembershipCommanderImpl @Inject() (
...
+    } else if (!userCanJoinLibraryWithoutInvite && inviteList.isEmpty && existingActiveMembership.isEmpty) {
+      // private library & no library invites with matching authtoken
+      Left(LibraryFail(FORBIDDEN, "cant_join_library_without_an_invite"))
+    } else {
+      val maxAccess: LibraryAccess = {
+        val orgMemberAccess: Option[LibraryAccess] = if (lib.isSecret) None else lib.organizationId.flatMap { orgId =>
+          lib.organizationMemberAccess.filter { _ => organizationMembershipCommander.getMembership(orgId, userId).isDefined }
+        }
+        (inviteList.map(_.access).toSet ++ orgMemberAccess).maxOpt.getOrElse(LibraryAccess.READ_ONLY)
+      }
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryMembershipCommander.scala:145]: @andrewconner @eishay

File: kifi-backend/modules/shoebox/app/com/keepit/model/LibraryRepo.scala [hunk 5]
```diff
@@ -227,6 +230,12 @@ class LibraryRepoImpl @Inject() (
     q.list
   }
 
+  def getLibrariesWithOpenWriteAccess(organizationId: Id[Organization], excludeState: Option[State[Library]] = Some(LibraryStates.INACTIVE))(implicit session: RSession): Seq[Library] = {
+    val collaborativePermissions = LibraryAccess.collaborativePermissions
+    val q = for { r <- rows if r.orgId === organizationId && r.state =!= excludeState.orNull && r.orgMemberAccess.inSet(collaborativePermissions) } yield r
+    q.list
+  }
+
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/LibraryRepo.scala:238]: @andrewconner @ryanpbrewster

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala [hunk 6]
```diff
@@ -44,7 +43,8 @@ case class Library(
     lastKept: Option[DateTime] = None,
     keepCount: Int = 0,
     whoCanInvite: Option[LibraryInvitePermissions] = None,
-    organizationId: Option[Id[Organization]] = None) extends ModelWithPublicId[Library] with ModelWithState[Library] with ModelWithSeqNumber[Library] {
+    organizationId: Option[Id[Organization]] = None,
+    organizationMemberAccess: Option[LibraryAccess] = None) extends ModelWithPublicId[Library] with ModelWithState[Library] with ModelWithSeqNumber[Library] {
```
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala:47]: @andrewconner @ryanpbrewster

File: kifi-backend/modules/common/app/com/keepit/model/OrganizationFactory.scala [hunk 7]
```diff
@@ -31,7 +31,7 @@ object OrganizationFactory {
     def withMembers(newMembers: Seq[User]) = this.copy(members = newMembers)
     def withInvitedUsers(newInvitedUsers: Seq[User]) = this.copy(invitedUsers = newInvitedUsers)
     def withInvitedEmails(newInvitedEmails: Seq[EmailAddress]) = this.copy(invitedEmails = newInvitedEmails)
-    def withHandle(newHandle: OrganizationHandle) = this.copy(org = org.copy(primaryHandle = Some(PrimaryOrganizationHandle(newHandle, newHandle))))
+    def withHandle(newHandle: OrganizationHandle): PartialOrganization = this.copy(org = org.copy(primaryHandle = Some(PrimaryOrganizationHandle(newHandle, newHandle))))
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/OrganizationFactory.scala:34]: Do we have `org.withHandle` defined?

Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/model/OrganizationFactory.scala:34]: Nope I think the normalization happens in `HandleCommander`

**Comment by ryanpbrewster** [comment]
Big PR. Most of it looks good to me. Hard to tell what stuff was just moved and what stuff changed. I'll take a closer look tomorrow morning.

File: kifi-backend/modules/shoebox/marketing (hunk 3 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/marketing:1]: Go into the directory and `git pull origin master` to grab the latest code. If that doesn't bring you to the correct commit hash, do `git checkout 207f54c6f63ecc7f8144aab5acb1aefb0b4da105`.

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala (hunk 6 above)
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala:47]: How about no default param (for this or org). Keep it explicit?

Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala:47]: How about no default param (for this or org). Keep it explicit?

**Comment by andrewconner** [comment]
:+1:

File: kifi-backend/modules/shoebox/conf/shoeboxService.routes (hunk 2 above)
Comment (adamjgrant) [inline kifi-backend/modules/shoebox/conf/shoeboxService.routes:115]: (Making sure I'm reading this correctly) By default, this route will not include org libraries?

Comment (adamjgrant) [inline kifi-backend/modules/shoebox/conf/shoeboxService.routes:115]: @cvializ We'd probably need to explicitly pass in `true` to this route for the extension. I don't think we do this today, do we?

Comment (andrewconner) [inline kifi-backend/modules/shoebox/conf/shoeboxService.routes:115]: API compatibility flag. Next ext version will include this set to true.

Comment (cvializ) [inline kifi-backend/modules/shoebox/conf/shoeboxService.routes:115]: lgtm

**Comment by ryanpbrewster** [comment]
I think we need to be really careful with this PR. We have a lot of implicit assumptions all over the place that are really hard to catch because of the implicit library permissions model.

Things like `LibraryWriteAction` need to be updated so that controllers don't bail immediately.

**Comment by ryanpbrewster** [comment]
Nevermind, apparently frontend will do some of the lifting here. Ignore me.

**Comment by ryanpbrewster** [comment]
This lgtm. I'd feel better if you added a test that made sure that
1. Libraries with this permission set allow org members to do the right thing
2. Libraries without it don't let people do the wrong thing

**Pull request merged**